osh/builtin_trap.py

```diff
@@ -83,6 +55,96 @@ def RemoveUserHook(self, hook_name):
     mylib.dict_remove(self.hooks, hook_name)
 
 
+class SignalState(object):
+  """All changes to global signal state go through this object."""
+
+  def __init__(self):
+    # type: () -> None
+    self.display = None  # type: _IDisplay
+    # signal/hook name -> handler
+    self.traps = {}  # type: Dict[int, command_t]
+    # appended to by signal handlers
+    self.nodes_to_run = []  # type: List[command_t]
+
+  def InitShell(self):
+    # type: () -> None
+    """Always called when initializing the shell process."""
+    pyos.InitShell()
+
+  def InitInteractiveShell(self, display, my_pid):
+    # type: (_IDisplay, int) -> None
+    """Called when initializing an interactive shell."""
+    # The shell itself should ignore Ctrl-\.
+    pyos.Sigaction(SIGQUIT, SIG_IGN)
+
+    # This prevents Ctrl-Z from suspending OSH in interactive mode.
+    pyos.Sigaction(SIGTSTP, SIG_IGN)
+
+    # More signals from
+    # https://www.gnu.org/software/libc/manual/html_node/Initializing-the-Shell.html
+    # (but not SIGCHLD)
+    pyos.Sigaction(SIGTTOU, SIG_IGN)
+    pyos.Sigaction(SIGTTIN, SIG_IGN)
+
+    # Register a callback to receive terminal width changes.
+    # NOTE: In line_input.c, we turned off rl_catch_sigwinch.
+
+    # This is ALWAYS on, which means that it can cause EINTR, and wait() and
+    # read() have to handle it
+    self.display = display
+    pyos.RegisterSignalInterest(SIGWINCH)
+    pyos.SetSigwinchCode(pyos.UNTRAPPED_SIGWINCH)
+
+  def GetLastSignal(self):
+    # type: () -> int
+    """Return the last signal that fired"""
+    return pyos.LastSignal()
+
+  def AddUserTrap(self, sig_num, handler):
+    # type: (int, command_t) -> None
+    """For user-defined handlers registered with the 'trap' builtin."""
+
+    if sig_num == SIGWINCH:
+      assert self.display is not None
+      pyos.SetSigwinchCode(SIGWINCH)
+    else:
+      pyos.RegisterSignalInterest(sig_num)
+    self.traps[sig_num] = handler
+    # TODO: SIGINT is similar: set a flag, then optionally call user _TrapHandler
+
+  def RemoveUserTrap(self, sig_num):
+    # type: (int) -> None
+    """For user-defined handlers registered with the 'trap' builtin."""
+    # Restore default
+    mylib.dict_remove(self.traps, sig_num)
+
+    if sig_num == SIGWINCH:
+      pyos.SetSigwinchCode(pyos.UNTRAPPED_SIGWINCH)
+    else:
+      pyos.Sigaction(sig_num, SIG_DFL)
+    # TODO: SIGINT is similar: set a flag, then optionally call user _TrapHandler
+
+  def TakeRunList(self):
+      # type: () -> List[command_t]
+      """Transfer ownership of the current queue of pending trap handlers to the caller."""
+      sig_queue = pyos.GetPendingSignals()
+
+      run_list = []  # type: List[command_t]
+      for sig_num in sig_queue:
+        node = self.traps.get(sig_num, None)
```

Comment:
FWIW it looks like self.traps is no longer read (or written) by the signal handler

So my instinct is to move it out of here, but it's not a big deal

i.e. I was thinking the ONLY members of the `SignalSafe` / `SignalShared` are the ones that can potentially be raced on

I guess one place to put it is to unify it with HookState (could be back to TrapState LOL)

Reply:
Agh this one is awkward. If we move `SignalState.traps` over to `HookState` (or `TrapState` :smile:), `SignalState`'s member functions will basically just be passthrough to `pyos`... Then `TakeRunList()` now either needs to move too or `SignalState` needs to have a `HookState` as a member so it can find the `command_t`s for each signal. The semantics of `AddUserTrap()` would also get kind of weird. What if we combine `SignalState` and `HookState` into `TrapState` and relegate all the racey bits to `pyos` (where they already live)?

Reply:
I was actually thinking that we can have one class with racey shared state (self.run_list) and then one class with the main thread state (self.hooks, self.traps)

But I think that will be clearer when we do the TSAN thing (which I suppose is not that urgent, since this looks pretty clean to me)

So this seems good for now